newapi/ooniapi/auth.py

```diff
@@ -0,0 +1,465 @@
+    - call the citizenlab CRUD entry points using the cookie
+    - call bookmarked searches/urls/msmts entry points using the cookie
+
+Configuration parameters:
+    BASE_URL
+    JWT_ENCRYPTION_KEY
+    MAIL_SERVER
+    MAIL_PORT
+    MAIL_USERNAME
+    MAIL_PASSWORD
+    MAIL_USE_SSL
+    MAIL_SOURCE_ADDRESS
+    LOGIN_EXPIRY_DAYS
+    SESSION_EXPIRY_DAYS
+"""
+
+# Courtesy of https://emailregex.com/
+EMAIL_RE = re.compile(r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)")
+
+
+def jerror(msg, code=400):
+    return make_response(jsonify(error=msg), code)
+
+
+def create_jwt(payload: dict) -> str:
+    key = current_app.config["JWT_ENCRYPTION_KEY"]
+    return jwt.encode(payload, key, algorithm="HS256").decode()
+
+
+def decode_jwt(token, **kw):
+    key = current_app.config["JWT_ENCRYPTION_KEY"]
+    return jwt.decode(token, key, algorithms=["HS256"], **kw)
+
+
+def hash_email_address(email_address: str) -> str:
+    key = current_app.config["JWT_ENCRYPTION_KEY"].encode()
+    em = email_address.encode()
+    return hashlib.blake2b(em, key=key, digest_size=16).hexdigest()
+
+
+def role_required(roles):
+    # Decorator requiring user to be logged in and have the right role
+    # Also refreshes the session
+    if isinstance(roles, str):
+        roles = [roles]
+
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            token = request.cookies.get("ooni", "")
+            try:
+                tok = decode_jwt(token, audience="user_auth")
+                del token
+                if tok["role"] not in roles:
+                    return jerror("Role not authorized", 401)
+            except Exception:
+                return jerror("Authentication required", 401)
+
+            # check for session expunge
+            # TODO: cache query
+            query = """SELECT threshold
+                FROM session_expunge
+                WHERE account_id = :account_id """
+            account_id = tok["account_id"]
+            query_params = dict(account_id=account_id)
+            q = current_app.db_session.execute(query, query_params)
+            row = q.fetchone()
+            if row:
+                iat = datetime.utcfromtimestamp(tok["iat"])
+                threshold = row[0]
+                if iat < threshold:
+                    return jerror("Authentication token expired", 401)
+
+            # attach nickname to request
+            request._user_nickname = tok["nick"]
+            # run the HTTP route method
+            resp = func(*args, **kwargs)
+
+            token_age = time.time() - tok["iat"]
+            if token_age > 600:  # refresh token if needed
+                newtoken = _create_session_token(
+                    tok["account_id"], tok["nick"], tok["role"], tok["login_time"]
+                )
+                resp.set_cookie(
+                    "ooni", newtoken, secure=True, httponly=True, samesite="Strict"
```

Comment:
Perhaps this can be made into a function for setting the cookie so that we can be sure we always set the cookies in the same way?

Reply:
Done